frontend/ui/trapper.lua

```diff
@@ -0,0 +1,133 @@
+-- and use it as a wrapper.
+--
+-- @cmd text shell cmd to execute and get output from
+-- @infomessage ui.widget.InfoMessage already shown InfoMessage widget
+-- @check_interval_sec[opt=0.1] float interval in second for checking pipe for available output
+-- @treturn bool completed (true if not interrupted, false if dismissed)
+-- @treturn string output of command
+function Trapper:dismissablePopen(cmd, infomessage, check_interval_sec)
+
+    -- We need to be called from an already created and running coroutine.
+    -- The first yield() we'll do will return to where the caller called
+    -- us, where it should probably just end the processing of some event
+    -- and give control back to UIManager (UIManager should then execute a
+    -- scheduled coroutine.resume(true), or process a onTapClose on our
+    -- InfoMessage that will do do a coroutine.resume(false) ).
+    local _coroutine = coroutine.running()
+    assert(_coroutine ~= nil, "Need to be called from a coroutine")
+
+    -- We need an InfoMessage, so we can plug a dismiss_callback to it
+    -- so that it can be used to abort the waiting for cmd output.
+    assert(infomessage ~= nil, "Need an infomessage to plug a dismiss callback to")
+    infomessage.dismiss_callback = function()
+        -- this callback will resume us at coroutine.yield() below
+        -- with a go_on = false
+        coroutine.resume(_coroutine, false)
+    end
+
+    if not check_interval_sec then
+        check_interval_sec = 0.1 -- default: check for output every 100ms
+    end
+    local collect_interval_sec = 5 -- collect cancelled cmd every 5 second, no hurry
+    local completed = false
+    local output = nil
+
+    local std_out = io.popen(cmd, "r")
+    if std_out then
+        -- We check regularly if data is available to be read, and we give
+        -- control in the meantime to UIManager so our InfoMessage.dismiss_callback
+        -- get a chance to be triggered, in which case we won't wait for reading,
+        -- We'll schedule a background function to collect the uneeded output later and
+        -- close the pipe.
+        while true do
+            -- The following function will resume us at coroutine.yield() below
+            -- with a go_on = true
+            local go_on_func = function() coroutine.resume(_coroutine, true) end
+            UIManager:scheduleIn(check_interval_sec, go_on_func) -- check every 100ms
+            local go_on = coroutine.yield() -- Gives control back to UIManager
+            if not go_on then -- the dismiss_callback resumed us
+                UIManager:unschedule(go_on_func)
+                -- We forget cmd here, but something has to collect
+                -- its output and close the pipe to not leak file handles and
+                -- zombie processes.
+                local collect_and_clean
+                collect_and_clean = function()
+                    if ffiutil.getNonBlockingReadSize(std_out) ~= 0 then -- cmd started outputing
```

Comment:
We're not worried about that nil/-1 thing?

Reply:
Nope, no more :)
If unsupported, it returns now nil (no more -1), and so ~= 0 would make it go on with reading (and block, as before, we can't do any better).